### PR TITLE
[Fixes #7115] Update property type check in SaveTempDataPropertyFilte…

### LINKED ISSUE #7115: Update property type check in SaveTempDataPropertyFilterBase to match TempDataSerializer requirements
Follow up to #6591 - I was working on a template change where I was passing around a `string[]` via TempData. This works fine because the serializer supports arrays of primitives. However, applying `TempDataAttribute` on a property fails saying the property is not a primitive or string type. We should either update the check in the property to match the serializer or remove it entirely and let the serialized determine if it can correctly round trip a property.

### PR DESCRIPTION
**Pull request by kichalla**
…rBase to match TempDataSerializer requirements

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs [hunk 1]
```diff
@@ -98,10 +98,18 @@ private static void ValidateProperty(PropertyHelper propertyHelper)
...
+                var messageWithPropertyInfo = Resources.FormatTempDataProperties_InvalidType(
+                    property.DeclaringType.FullName,
+                    property.Name,
+                    nameof(TempDataAttribute));
+
+                errorMessage = errorMessage ?? Resources.FormatTempData_CannotSerializeType(
+                    typeof(TempDataSerializer).FullName,
+                    actualType);
+
+                throw new InvalidOperationException($"{messageWithPropertyInfo} {errorMessage}");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs]: The we we're building these error messages is just kind wacky. Use a stringbuilder with appendformat?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs [hunk 2]
```diff
@@ -98,10 +98,18 @@ private static void ValidateProperty(PropertyHelper propertyHelper)
 
             var propertyType = Nullable.GetUnderlyingType(property.PropertyType) ?? property.PropertyType;
 
-            if (!(propertyType.GetTypeInfo().IsPrimitive || propertyType == typeof(string)))
+            if (!TempDataSerializer.CanSerializeType(propertyType, out var actualType, out var errorMessage))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs]: Under what conditions will this return false with a null error message? Should we just solve that?

Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs]: Good point. I fixed this one now.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs [hunk 3]
```diff
@@ -98,10 +98,14 @@ private static void ValidateProperty(PropertyHelper propertyHelper)
...
             {
-                throw new InvalidOperationException(
-                    Resources.FormatTempDataProperties_PrimitiveTypeOrString(property.DeclaringType.FullName, property.Name, nameof(TempDataAttribute)));
+                var messageWithPropertyInfo = Resources.FormatTempDataProperties_InvalidType(
+                    property.DeclaringType.FullName,
+                    property.Name,
+                    nameof(TempDataAttribute));
+
+                throw new InvalidOperationException($"{messageWithPropertyInfo} {errorMessage}");
```
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs]: After this latest commit, do you still think we need a stringbuilder?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs]: nah this is OK now

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/TempDataSerializer.cs [hunk 4]
```diff
@@ -157,34 +157,50 @@ public byte[] Serialize(IDictionary<string, object> values)
 
         public void EnsureObjectCanBeSerialized(object item)
         {
+            item = item ?? throw new ArgumentNullException(nameof(item));
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/TempDataSerializer.cs]: 💯

**Pull request merged**